crypto/x509/by_store.c

```diff
@@ -111,11 +111,24 @@ static int by_store_ctrl_ex(X509_LOOKUP *ctx, int cmd, const char *argp,
 {
     switch (cmd) {
     case X509_L_ADD_STORE:
-        /* If no URI is given, use the default cert dir as default URI */
+        /* First try the newer default cert URI envvar. */
         if (argp == NULL)
+            argp = ossl_safe_getenv(X509_get_default_cert_uri_env());
+
+        /* If not set, see if we have a URI in the older cert dir envvar. */
+        if (argp == NULL) {
             argp = ossl_safe_getenv(X509_get_default_cert_dir_env());
+            if (argp != NULL && !ossl_is_uri(argp))
```

Comment:
If we keep the check for `://` in ossl_is_uri IMO we should NOT do this check here as I said that URIs are definitely valid without `//`.

Reply:
What currently happens in 3.0 is:

Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`, search store `foo:bar`
Specify `SSL_CERT_DIR=foo://bar` -> search dir `foo`, search dir `//bar`, search store `foo://bar`

So the question is how to fix this in a compatible way. Currently if someone specifies a store URI of `foo://bar`, the directory `foo` in the current directory is added as a root certificate directory which is quite alarming.

The possible fixes I see are:

Resolution A: If "://" is found in `SSL_CERT_DIR`, treat as an URI, else, as a list of paths
- URIs of form `scheme://` can use SSL_CERT_DIR, but URIs of form `scheme:` must use SSL_CERT_URI
- Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
- Pro: Fixes the ambiguity, which has potential security implications, for existing users
- Con: May cause breakage to anyone using a custom store with `foo:bar` URIs, who will need to use `SSL_CERT_URI` instead

Resolution B: If ":" is found in `SSL_CERT_DIR`, treat as an URI, else, as a single path
- Specify `SSL_CERT_DIR=foo` -> search dir `foo`
- Specify `SSL_CERT_DIR=foo:bar` -> search store `foo:bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
- Con: Very incompatible

Resolution C: keep the current behaviour for the existing environment variables and introduce two new environment variables
- Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar` store `foo:bar`
- Specify `SSL_CERT_DIR=foo://bar` -> search dir `foo`, search dir `//bar` store `foo://bar`
- `SSL_CERT_DIR` is deprecated but continues to work as before. The potential security issues are documented.
- A new environment variable `SSL_CERT_URI` becomes the preferred way to specify a URI
- A new environment variable `SSL_CERT_PATH` becomes the preferred way to specify a list of paths. Unlike `SSL_CERT_DIR` it is never processed as a URI.

Reply:
I prefer C.

I'd accept a minor variant of A, partially combined with C:

 - URIs of form `scheme://` can use SSL_CERT_DIR, but URIs of form `scheme:` should use SSL_CERT_URI
 - Specify `SSL_CERT_DIR=foo:bar` -> search dir `foo`, search dir `bar`, search store `foo:bar`
 - Specify `SSL_CERT_DIR=foo://bar` -> search store `foo://bar`
 - Pro: Partially fixes the ambiguity, which has potential security implications, for existing users
 - Con: Does not fix the ambiguity for `foo:bar` stores which means we would probably still need to deprecate SSL_CERT_DIR and introduce SSL_CERT_PATH

I am not sure the slight improvement of fixing the issue for existing users of `scheme://` with SSL_CERT_DIR is worth it.

Reply:
I'd prefer C as well (it resolves a naming bug as well.... Common nomenclature is that PATH indicates several concatenated directories while DIR is a single directory)